internal/service/meta/meta_service.go

```diff
@@ -92,3 +98,136 @@ func (ms *MetaService) GetMetaList(ctx context.Context, objID string) (metas []*
 	}
 	return metas, err
 }
+
+// GetReactionByObjectId get reaction
+func (ms *MetaService) GetReactionByObjectId(ctx context.Context, objectID string) (resp schema.ReactionResp, err error) {
+	reactionMeta, err := ms.GetMetaByObjectIdAndKey(ctx, objectID, entity.ObjectReactSummaryKey)
+
+	// if not exist, return nil
+	if err != nil {
+		var pacmanErr *myErrors.Error
+		if errors.As(err, &pacmanErr) && pacmanErr.Reason == reason.MetaObjectNotFound {
+			return nil, nil
+		} else {
+			return nil, err
+		}
+	}
+
+	var reaction schema.ReactSummaryMeta
+	err = json.Unmarshal([]byte(reactionMeta.Value), &reaction)
+	if err != nil {
+		return nil, err
+	}
+	return ms.convertToReactionResp(ctx, reaction)
+}
+
+// AddOrUpdateReaction add or update reaction
+func (ms *MetaService) AddOrUpdateReaction(ctx context.Context, req *schema.UpdateReactionReq) (resp schema.ReactionResp, err error) {
+	// get reaction for this object
+	reactionMeta, err := ms.GetMetaByObjectIdAndKey(ctx, req.ObjectID, entity.ObjectReactSummaryKey)
+
+	var reaction schema.ReactSummaryMeta
+	if err != nil {
+		var pacmanErr *myErrors.Error
+		if errors.As(err, &pacmanErr) && pacmanErr.Reason == reason.MetaObjectNotFound {
+			// create new reaction summary
+			reaction = schema.ReactSummaryMeta{}
+		} else {
+			return nil, err
+		}
+	} else {
+		// json unmarshal reactionMeta.Value to reaction
+		err = json.Unmarshal([]byte(reactionMeta.Value), &reaction)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// update reaction
+	ms.updateReaction(req, reaction)
+
+	// write back to meta repo
+	reactSumBytes, err := json.Marshal(reaction)
+	if err != nil {
+		return nil, err
+	}
+
+	if reactionMeta == nil {
+		err = ms.AddMeta(ctx, req.ObjectID, entity.ObjectReactSummaryKey, string(reactSumBytes))
+	} else {
+		err = ms.UpdateMeta(ctx, reactionMeta.ID, entity.ObjectReactSummaryKey, string(reactSumBytes))
+	}
```

Comment:
Thanks for your suggestion! Updated to use Transaction and refactored these part. Also using ForUpdate to avoid override when concurrency. Pls help to review and welcome more comments.